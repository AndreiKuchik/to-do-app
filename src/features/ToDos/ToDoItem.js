import React from 'react';
import PropTypes from 'prop-types';
import { deleteToDo, completeToDo } from '../toDoListSlice';
import { useAppDispatch } from '../../core/hooks';

import styles from './ToDoItem.module.css';

const ToDoItem = (props) => {
  const dispatch = useAppDispatch();

  return (
    <React.Fragment>
      <input
        role="todo-item-checkbox"
        type="checkbox"
        checked={props.isCompleted}
        onChange={() => dispatch(completeToDo(props.id))}
        className={styles.isCompletedCheckBox}
      />
      <div
        role="todo-item-text"
        className={`${styles.toDoText} ${props.isCompleted && styles.crossed}`}
      >
        {props.text}
      </div>
      <button
        type="button"
        role="todo-item-delete-button"
        className={styles.deleteButton}
        onClick={() => dispatch(deleteToDo(props.id))}
      >
        Delete
      </button>
    </React.Fragment>
  );
};

ToDoItem.propTypes = {
  id: PropTypes.string.isRequired,
  text: PropTypes.string.isRequired,
  isCompleted: PropTypes.bool.isRequired,
};

export default ToDoItem;
