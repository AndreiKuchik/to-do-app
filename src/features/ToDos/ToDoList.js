/* eslint-disable no-shadow */
import React from 'react';
import { DragDropContext, Droppable, Draggable } from 'react-beautiful-dnd';
import { useAppDispatch, useAppSelector } from '../../core/hooks';
import Card from '../../components/Card';
import { selectToDoList, reorderList } from '../toDoListSlice';
import styles from './ToDoList.module.css';
import ToDoItem from './ToDoItem';

const ToDoList = () => {
  const dispatch = useAppDispatch();
  const toDoList = useAppSelector(selectToDoList);

  const onDragEnd = (result) => {
    const dragElementId = toDoList[result.source.index].id;
    const dropElementId = toDoList[result.destination.index].id;
    const dragAndDropAction = {
      dragElementId,
      dropElementId,
    };
    dispatch(reorderList(dragAndDropAction));
  };

  return (
    <Card className={styles.toDoList}>
      {toDoList.length !== 0 && (
        <DragDropContext onDragEnd={onDragEnd}>
          <Droppable droppableId="toDoList">
            {(provided) => (
              <ul {...provided.droppableProps} ref={provided.innerRef}>
                {toDoList.map((todo, index) => (
                  <Draggable key={todo.id} draggableId={todo.id} index={index}>
                    {(provided) => (
                      <li
                        ref={provided.innerRef}
                        {...provided.draggableProps}
                        {...provided.dragHandleProps}
                      >
                        <ToDoItem
                          key={todo.id}
                          id={todo.id}
                          text={todo.toDo}
                          isCompleted={todo.isCompleted}
                        />
                      </li>
                    )}
                  </Draggable>
                ))}
                {provided.placeholder}
              </ul>
            )}
          </Droppable>
        </DragDropContext>
      )}
      {toDoList.length === 0 && (
        <h3 role="empty-list-text">ToDo list is empty</h3>
      )}
    </Card>
  );
};

export default ToDoList;
