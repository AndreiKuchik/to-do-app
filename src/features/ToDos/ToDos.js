import React, { useEffect } from 'react';
import { useAppDispatch, useAppSelector } from '../../core/hooks';
import Card from '../../components/Card';
import ToDoFooter from '../ToDoFooter/ToDoFooter';
import { fetchToDoList } from '../toDosSlice';
import styles from './ToDos.module.css';
import ToDoList from './ToDoList';
import Statuses from '../../core/statusesEnum';
import ErrorMesssage from '../../components/ErrorMessage';
import LoadingSpinner from '../../components/LoadingSpinner';

const ToDos = () => {
  const dispatch = useAppDispatch();
  const listStatus = useAppSelector((state) => state.toDoList.status);

  useEffect(() => {
    if (listStatus === Statuses.Undefined) {
      dispatch(fetchToDoList());
    }
  }, [listStatus, dispatch]);

  let content;

  if (listStatus === Statuses.Loading) {
    content = <LoadingSpinner />;
  } else if (listStatus === Statuses.Succeeded) {
    content = (
      <React.Fragment>
        <ToDoList />
        <ToDoFooter />
      </React.Fragment>
    );
  } else if (listStatus === Statuses.Failed) {
    content = <ErrorMesssage />;
  }

  return <Card className={styles.toDos}>{content}</Card>;
};

export default ToDos;
