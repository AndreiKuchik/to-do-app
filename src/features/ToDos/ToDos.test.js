import React from 'react';
import { render, screen } from '@testing-library/react';
import { Provider } from 'react-redux';
import configureStore from 'redux-mock-store';
import ToDos from './ToDos';
import Statuses from '../../common/requestStatusesEnum';
import Filters from '../../common/toDoFiltersEnum';

const mockStore = configureStore([]);

describe('Test ToDos component', () => {
  let store;

  it('should match with snapshot when status is `loading`', () => {
    store = mockStore({
      toDoList: {
        status: Statuses.Loading,
      },
    });

    const { container } = render(
      <Provider store={store}>
        <ToDos />
      </Provider>,
    );

    expect(container).toMatchSnapshot();
  });

  it('should match with snapshot when status is `succeeded`', () => {
    store = mockStore({
      toDoList: {
        toDoList: [
          {
            id: '1',
            toDo: 'ToDo1',
            isCompleted: false,
          },
          {
            id: '2',
            toDo: 'ToDo2',
            isCompleted: true,
          },
        ],
        status: Statuses.Succeeded,
        filter: Filters.All,
      },
    });

    const { container } = render(
      <Provider store={store}>
        <ToDos />
      </Provider>,
    );

    expect(container).toMatchSnapshot();
  });

  it('should match with snapshot when status is `failed`', () => {
    store = mockStore({
      toDoList: {
        status: Statuses.Failed,
      },
    });

    const { container } = render(
      <Provider store={store}>
        <ToDos />
      </Provider>,
    );
    expect(container).toMatchSnapshot();
  });

  it('error message should be displayed when status is `failed`', () => {
    store = mockStore({
      toDoList: {
        status: Statuses.Failed,
      },
    });

    render(
      <Provider store={store}>
        <ToDos />
      </Provider>,
    );

    const errorMessage = screen.getByText(
      'Something went wrong. Could you please contact administrator.',
    );
    expect(errorMessage).toBeInTheDocument();
  });
  it('loading message should be displayed when status is `loading`', () => {
    store = mockStore({
      toDoList: {
        status: Statuses.Loading,
      },
    });

    render(
      <Provider store={store}>
        <ToDos />
      </Provider>,
    );

    expect(screen.getByRole('loading-spinner')).toBeInTheDocument();
  });
});
