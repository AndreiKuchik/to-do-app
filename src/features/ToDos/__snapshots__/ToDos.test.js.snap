// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Test ToDos component should match with snapshot when status is \`failed\` 1`] = `
<div>
  <div
    class="card toDos"
  >
    <div
      class="errorMessage"
    >
      Something went wrong. Could you please contact administrator.
    </div>
  </div>
</div>
`;

exports[`Test ToDos component should match with snapshot when status is \`loading\` 1`] = `
<div>
  <div
    class="card toDos"
  >
    <p
      class="loading"
      role="loading-spinner"
    >
      <span
        style="display: inherit; position: relative; font-size: 0px; height: 50px; width: 50px;"
      >
        <span
          style="width: 0px; height: 0px; border-right: 25px solid transparent; border-top: 25px solid transparent; border-left: 25px solid #36d7b7; border-bottom: 25px solid #36d7b7; border-radius: 25px; position: absolute; animation: react-spinners-PacmanLoader-pacman-1 0.8s infinite ease-in-out; animation-fill-mode: both;"
        />
        <span
          style="width: 0px; height: 0px; border-right: 25px solid transparent; border-top: 25px solid #36d7b7; border-left: 25px solid #36d7b7; border-bottom: 25px solid transparent; border-radius: 25px; position: absolute; animation: react-spinners-PacmanLoader-pacman-2 0.8s infinite ease-in-out; animation-fill-mode: both;"
        />
        <span
          style="width: 8.333333333333334px; height: 8.333333333333334px; background-color: rgb(54, 215, 183); margin: 2px; border-radius: 100%; transform: translate(0, -6.25px); position: absolute; top: 25px; left: 100px; animation: react-spinners-PacmanLoader-ball 1s 0.5s infinite linear; animation-fill-mode: both;"
        />
        <span
          style="width: 8.333333333333334px; height: 8.333333333333334px; background-color: rgb(54, 215, 183); margin: 2px; border-radius: 100%; transform: translate(0, -6.25px); position: absolute; top: 25px; left: 100px; animation: react-spinners-PacmanLoader-ball 1s 0.75s infinite linear; animation-fill-mode: both;"
        />
        <span
          style="width: 8.333333333333334px; height: 8.333333333333334px; background-color: rgb(54, 215, 183); margin: 2px; border-radius: 100%; transform: translate(0, -6.25px); position: absolute; top: 25px; left: 100px; animation: react-spinners-PacmanLoader-ball 1s 1s infinite linear; animation-fill-mode: both;"
        />
        <span
          style="width: 8.333333333333334px; height: 8.333333333333334px; background-color: rgb(54, 215, 183); margin: 2px; border-radius: 100%; transform: translate(0, -6.25px); position: absolute; top: 25px; left: 100px; animation: react-spinners-PacmanLoader-ball 1s 1.25s infinite linear; animation-fill-mode: both;"
        />
      </span>
    </p>
  </div>
</div>
`;

exports[`Test ToDos component should match with snapshot when status is \`succeeded\` 1`] = `
<div>
  <div
    class="card toDos"
  >
    <div
      class="card toDoList"
    >
      <ul
        data-rbd-droppable-context-id="0"
        data-rbd-droppable-id="toDoList"
      >
        <li
          aria-describedby="rbd-hidden-text-0-hidden-text-0"
          data-rbd-drag-handle-context-id="0"
          data-rbd-drag-handle-draggable-id="1"
          data-rbd-draggable-context-id="0"
          data-rbd-draggable-id="1"
          draggable="false"
          role="button"
          tabindex="0"
        >
          <input
            class="isCompletedCheckBox"
            role="todo-item-checkbox"
            type="checkbox"
          />
          <div
            class="toDoText false"
            role="todo-item-text"
          >
            ToDo1
          </div>
          <button
            class="deleteButton"
            role="todo-item-delete-button"
            type="button"
          >
            Delete
          </button>
        </li>
        <li
          aria-describedby="rbd-hidden-text-0-hidden-text-0"
          data-rbd-drag-handle-context-id="0"
          data-rbd-drag-handle-draggable-id="2"
          data-rbd-draggable-context-id="0"
          data-rbd-draggable-id="2"
          draggable="false"
          role="button"
          tabindex="0"
        >
          <input
            checked=""
            class="isCompletedCheckBox"
            role="todo-item-checkbox"
            type="checkbox"
          />
          <div
            class="toDoText crossed"
            role="todo-item-text"
          >
            ToDo2
          </div>
          <button
            class="deleteButton"
            role="todo-item-delete-button"
            type="button"
          >
            Delete
          </button>
        </li>
      </ul>
    </div>
    <div
      class="footer"
    >
      <span
        role="to-do-item-amount"
      >
        2
         Item left
      </span>
      <div>
        <button
          class="filterButton"
          type="button"
        >
          All
        </button>
        <button
          class="filterButton"
          role="active-filter-button"
          type="button"
        >
          Active
        </button>
        <button
          class="filterButton"
          role="completed-filter-button"
          type="button"
        >
          Completed
        </button>
      </div>
      <button
        class="completeAllButton"
        role="clear-completed-button"
        type="button"
      >
        Clear Completed
      </button>
    </div>
  </div>
</div>
`;
